return (function(game)
  local Services = {
    workspace = game:GetService("Workspace"),
    coregui = game:GetService("CoreGui"),
    players = game:GetService("Players"),
    lighting = game:GetService("Lighting"),
    replicatedstorage = game:GetService("ReplicatedStorage"),
    serverstorage = game:GetService("ServerStorage"),
    serverscriptservice = game:GetService("ServerScriptService"),
    startergui = game:GetService("StarterGui"),
    starterpack = game:GetService("StarterPack"),
    starterplayers = game:GetService("StarterPlayer"),
    soundservice = game:GetService("SoundService"),
    textservice = game:GetService("TextService"),
    tweenservice = game:GetService("TweenService"),
    uis = game:GetService("UserInputService"),
    runservice = game:GetService("RunService"),
    hapticservice = game:GetService("HapticService"),
    vrservice = game:GetService("VRService"),
    pathfindingservice = game:GetService("PathfindingService"),
    physicsservice = game:GetService("PhysicsService"),
    collectionservice = game:GetService("CollectionService"),
    httpservice = game:GetService("HttpService"),
    teleportservice = game:GetService("TeleportService"),
    marketplace = game:GetService("MarketplaceService"),
    chat = game:GetService("Chat"),
    datastoreservice = game:GetService("DataStoreService"),
    contextactionservice = game:GetService("ContextActionService"),
    policyservice = game:GetService("PolicyService"),
    teams = game:GetService("Teams"),
    badgeservice = game:GetService("BadgeService"),
    localizationservice = game:GetService("LocalizationService"),
    analyticsservice = game:GetService("AnalyticsService"),
    debris = game:GetService("Debris"),
    messagingservice = game:GetService("MessagingService"),
    guiservice = game:GetService("GuiService"),
  }

  return Services
end)